File: packages/retrieval/eval/longmemeval/adapter.ts
import { decode, encode } from 'gpt-tokenizer';
import type { UpsertEntry, QueryHit } from '../../src/index';
import type { ChunkMode, LmeRecord, LmeSession } from './types';

// text-embedding-3-small accepts at most 8191 tokens per input. A char-based cap
// can't bound tokens safely (some characters encode to multiple BPE tokens), so
// budget by actual token count (cl100k_base, the model's encoding). Keep a wide
// margin under the hard limit: our local count can run slightly under OpenAI's
// own tokenizer, and decode/re-encode at split boundaries can drift, so cap well
// below 8191 (the embedder self-heals any residual over-long input). Every part
// keeps the session's session_id, so recall (which matches on session_id) is
// unaffected.
const MAX_EMBED_TOKENS = 6000;

// Conversation text can contain literal special-token strings (e.g.
// "<|endoftext|>"); encode them as ordinary text rather than throwing, matching
// how the embeddings API treats raw input.
const ENCODE_OPTS = { disallowedSpecial: new Set<string>() };

/** Token count of `text` under the embedder's encoding (cl100k_base). */
function tokenLen(text: string): number {
  return encode(text, ENCODE_OPTS).length;
}

/** Hard-slice a string into consecutive pieces each at most `budget` tokens. */
function sliceToBudget(text: string, budget: number): string[] {
  const tokens = encode(text, ENCODE_OPTS);
  if (tokens.length <= budget) return [text];
  const parts: string[] = [];
  for (let i = 0; i < tokens.length; i += budget) {
    parts.push(decode(tokens.slice(i, i + budget)));
  }
  return parts;
}

/** Pack a session's turns into newline-joined chunks each within `budget` tokens. */
function packTurns(session: LmeSession, budget: number): string[] {
  const lines: string[] = [];
  for (const turn of session) {
    const line = `${turn.role}: ${turn.content}`;
    if (tokenLen(line) <= budget) {
      lines.push(line);
    } else {
      // A single turn larger than the budget is hard-sliced so it still embeds.
      lines.push(...sliceToBudget(line, budget));
    }
  }
  const chunks: string[] = [];
  let current = '';
  for (const line of lines) {
    if (current.length > 0 && tokenLen(`${current}\n${line}`) > budget) {
      chunks.push(current);
      current = line;
    } else {
      current = current.length === 0 ? line : `${current}\n${line}`;
    }
  }
  if (current.length > 0) chunks.push(current);
  return chunks;
}

/**
 * Turn a LongMemEval haystack into UpsertEntry chunks.
 * - 'session' (default): one chunk per session (turns joined); sessions longer
 *   than the embedder's input budget are split into multiple chunks that all
 *   carry the same session_id.
 * - 'turn': one chunk per turn.
 * The id encodes the session index (+ turn/part index when split); fields carry
 * the session_id tag (+ date tag when present) so recall can match evidence.
 */
export function chunkRecord(record: LmeRecord, mode: ChunkMode): UpsertEntry[] {
  const entries: UpsertEntry[] = [];
  record.haystack_sessions.forEach((session, sIdx) => {
    const sessionId = record.haystack_session_ids[sIdx] ?? `session_${sIdx}`;
    const date = record.haystack_dates?.[sIdx];
    const baseFields: Record<string, string> = { session_id: sessionId };
    if (date !== undefined && date !== '') {
      baseFields.date = date;
    }

    if (mode === 'turn') {
      session.forEach((turn, tIdx) => {
        const text = `${turn.role}: ${turn.content}`;
        // A single turn can exceed the embedder budget too; hard-slice it like
        // session mode so it still embeds instead of failing the chunk.
        const parts = sliceToBudget(text, MAX_EMBED_TOKENS);
        parts.forEach((part, pIdx) => {
          entries.push({
            id: parts.length === 1 ? `s${sIdx}_t${tIdx}` : `s${sIdx}_t${tIdx}_p${pIdx}`,
            text: part,
            fields: { ...baseFields },
          });
        });
      });
    } else {
      const parts = packTurns(session, MAX_EMBED_TOKENS);
      parts.forEach((text, pIdx) => {
        entries.push({
          id: parts.length === 1 ? `s${sIdx}` : `s${sIdx}_p${pIdx}`,
          text,
          fields: { ...baseFields },
        });
      });
    }
  });
  return entries;
}

/** A record is a recall HIT if any retrieved chunk's session_id is evidence. */
export function recordIsHit(hits: QueryHit[], answerSessionIds: string[]): boolean {
  const evidence = new Set(answerSessionIds);
  return hits.some((hit) => evidence.has(hit.fields.session_id));
}

File: packages/retrieval-py/eval/longmemeval/adapter.py
from __future__ import annotations

from betterdb_retrieval import QueryHit, UpsertEntry

from .types import ChunkMode, LmeRecord, LmeSession

# text-embedding-3-small accepts at most 8191 tokens per input. Cap each chunk
# well under that (~4 chars/token heuristic, with margin) so a long session is
# split into multiple chunks instead of failing the embedding call. Every part
# keeps the session's session_id, so recall (which matches on session_id) is
# unaffected.
MAX_EMBED_CHARS = 24000


def _slice_to_budget(text: str, budget: int) -> list[str]:
    """Hard-slice a string into consecutive pieces each at most ``budget`` chars."""
    return [text[i : i + budget] for i in range(0, len(text), budget)]


def _pack_turns(session: LmeSession, budget: int) -> list[str]:
    """Pack a session's turns into newline-joined chunks each within ``budget``."""
    lines: list[str] = []
    for turn in session:
        line = f"{turn['role']}: {turn['content']}"
        if len(line) <= budget:
            lines.append(line)
        else:
            # A single turn larger than the budget is hard-sliced so it still embeds.
            lines.extend(_slice_to_budget(line, budget))
    chunks: list[str] = []
    current = ""
    for line in lines:
        if len(current) > 0 and len(current) + 1 + len(line) > budget:
            chunks.append(current)
            current = line
        else:
            current = line if len(current) == 0 else f"{current}\n{line}"
    if len(current) > 0:
        chunks.append(current)
    return chunks


def chunk_record(record: LmeRecord, mode: ChunkMode) -> list[UpsertEntry]:
    """Turn a LongMemEval haystack into UpsertEntry chunks.

    - 'session' (default): one chunk per session (turns joined); sessions longer
      than the embedder's input budget are split into multiple chunks that all
      carry the same session_id.
    - 'turn': one chunk per turn.

    The id encodes the session index (+ turn/part index when split); fields carry
    the session_id tag (+ date tag when present) so recall can match evidence.
    """
    entries: list[UpsertEntry] = []
    session_ids = record["haystack_session_ids"]
    dates = record.get("haystack_dates")
    for s_idx, session in enumerate(record["haystack_sessions"]):
        session_id = session_ids[s_idx] if s_idx < len(session_ids) else f"session_{s_idx}"
        date = dates[s_idx] if dates is not None and s_idx < len(dates) else None
        base_fields: dict[str, str | int | float] = {"session_id": session_id}
        if date:
            base_fields["date"] = date

        if mode == "turn":
            for t_idx, turn in enumerate(session):
                text = f"{turn['role']}: {turn['content']}"
                # A single turn can exceed the embedder budget too; hard-slice it
                # like session mode so it still embeds instead of failing the chunk.
                parts = (
                    [text]
                    if len(text) <= MAX_EMBED_CHARS
                    else _slice_to_budget(text, MAX_EMBED_CHARS)
                )
                for p_idx, part in enumerate(parts):
                    entry_id = (
                        f"s{s_idx}_t{t_idx}" if len(parts) == 1 else f"s{s_idx}_t{t_idx}_p{p_idx}"
                    )
                    entries.append(UpsertEntry(id=entry_id, text=part, fields=dict(base_fields)))
        else:
            parts = _pack_turns(session, MAX_EMBED_CHARS)
            for p_idx, text in enumerate(parts):
                entry_id = f"s{s_idx}" if len(parts) == 1 else f"s{s_idx}_p{p_idx}"
                entries.append(UpsertEntry(id=entry_id, text=text, fields=dict(base_fields)))
    return entries


def record_is_hit(hits: list[QueryHit], answer_session_ids: list[str]) -> bool:
    """A record is a recall HIT if any retrieved chunk's session_id is evidence."""
    evidence = set(answer_session_ids)
    return any(hit.fields.get("session_id") in evidence for hit in hits)
